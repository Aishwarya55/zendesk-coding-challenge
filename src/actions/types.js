export const UPDATE_SUBSCRIPTION = "UPDATE_SUBSCRIPTION";

export const UPDATE_SUBSCRIPTION_SUCCESS = "UPDATE_SUBSCRIPTION_SUCCESS";

export const UPDATE_SUBSCRIPTION_FAILURE = "UPDATE_SUBSCRIPTION_FAILURE";

export const GET_CURRENT_SUBSCRIPTION = "GET_CURRENT_SUBSCRIPTION";

export const GET_CURRENT_SUBSCRIPTION_SUCCESS = "GET_CURRENT_SUBSCRIPTION_SUCCESS";

export const GET_CURRENT_SUBSCRIPTION_FAILURE = "GET_CURRENT_SUBSCRIPTION_FAILURE";

export const ASYNC_CALL_STARTED = "ASYNC_CALL_STARTED"

export const GET_PRICE_SUCCESS = "GET_PRICE_SUCCESS"

export const GET_PRICE_FAILURE = "GET_PRICE_FAILURE"

