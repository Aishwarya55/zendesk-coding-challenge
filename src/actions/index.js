import axios from "axios"
import BASE_URL from "../environment"
import {
    GET_PRICE_SUCCESS,
    GET_PRICE_FAILURE,
    GET_CURRENT_SUBSCRIPTION_SUCCESS,
    GET_CURRENT_SUBSCRIPTION_FAILURE,
    UPDATE_SUBSCRIPTION_FAILURE,
    UPDATE_SUBSCRIPTION_SUCCESS,
    ASYNC_CALL_STARTED
} from "./types"


axios.defaults.headers.common["Content-Type"]="application/json";


const ayncCallStarted = () => ({
    type: ASYNC_CALL_STARTED
  });

//Fetch Cost preview from Backend server
export const getPrice = (data) => async (dispatch) => {
    dispatch(ayncCallStarted())
    const response = await axios.post(`${BASE_URL}/preview`, data)

    if(response.status === 200){
        dispatch({
            type: GET_PRICE_SUCCESS,
            payload: response.data
        })
    }else{
        dispatch({
            type: GET_PRICE_FAILURE,
            payload: response.data
        })
    }
}

//Fetch Current Subscription from Backend server
export const getCurrentSubscription = () => async(dispatch) => {
    dispatch(ayncCallStarted())
    const response = await axios.get(`${BASE_URL}/current`)

    if(response.status === 200){
        dispatch({
            type: GET_CURRENT_SUBSCRIPTION_SUCCESS,
            payload: response.data
        })
    }else{
        dispatch({
            type: GET_CURRENT_SUBSCRIPTION_FAILURE,
            payload: response.data
        })
    }
}

//Update current subscription in backend server
export const updateSubscription = (data) => async(dispatch) => {
    dispatch(ayncCallStarted())
    const response = await axios.put(`${BASE_URL}/current`, data)

    if(response.status === 200){
        dispatch({
            type: UPDATE_SUBSCRIPTION_SUCCESS,
            payload: response.data
        })
    }else{
        dispatch({
            type: UPDATE_SUBSCRIPTION_FAILURE,
            payload: response.data
        })
    }

}
