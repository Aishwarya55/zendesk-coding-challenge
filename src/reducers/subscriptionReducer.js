import {
    GET_PRICE_SUCCESS,
    GET_PRICE_FAILURE,
    GET_CURRENT_SUBSCRIPTION_SUCCESS,
    UPDATE_SUBSCRIPTION_SUCCESS,
    UPDATE_SUBSCRIPTION_FAILURE,
    ASYNC_CALL_STARTED
} from "./../actions/types"


export const initialState = {
    loading: false,
    costPreview: {

        plan: "",
        name: "",
        seats: 0,
        cost: ""

    },
    currentSubscription: {
        plan: "basic",
        name: "Basic",
        seats: 0,
        cost: ""
    },
    previousSubscription: {
        plan: "basic",
        name: "Basic",
        seats: 0,
        cost: ""
    }
}


export default function subscriptionReducer(state = initialState, action) {

    switch (action.type) {
        case ASYNC_CALL_STARTED:
            return { ...state, loading: true}
        case GET_PRICE_SUCCESS:
            return { ...state, costPreview: action.payload, loading: false }
        case GET_PRICE_FAILURE:
            return { ...state }
        case GET_CURRENT_SUBSCRIPTION_SUCCESS:
            return { ...state, currentSubscription: action.payload, loading: false }
        case UPDATE_SUBSCRIPTION_SUCCESS:
            return {...state, previousSubscription: state.currentSubscription, currentSubscription: action.payload, loading: false}
        case UPDATE_SUBSCRIPTION_FAILURE:
            return {...state}
        default:
            return { ...state }
    }

}