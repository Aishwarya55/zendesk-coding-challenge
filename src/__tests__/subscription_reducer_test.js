import subscriptionReducer from "../reducers/subscriptionReducer"
import * as types from "../actions/types"
import {initialState} from "../reducers/subscriptionReducer"

describe('subscription reducer', () => {
    it('should return initial state', () => {
        expect(subscriptionReducer(undefined, {})).toMatchSnapshot()
    })

    it('should handle GET_PRICE_SUCCESS', () => {
        expect(subscriptionReducer({}, {
            type:types.GET_PRICE_SUCCESS,
            payload: {
                plan: "good",
                name: "Good",
                seats: "5",
                cost: 50
            }
        })).toEqual({
            costPreview: {
                plan: "good",
                name: "Good",
                seats: "5",
                cost: 50
            },
            loading: false
        })
    })

    it('it should handle UPDATE_SUBSCRIPTION_SUCCESS', () => {
        expect(subscriptionReducer({}, {
            type: types.UPDATE_SUBSCRIPTION_SUCCESS,
            payload: {
                plan: "best",
                name: "Best",
                seats: "5",
                cost: 5000
            }
        })).toEqual({
       
            currentSubscription: {
                plan: "best",
                name: "Best",
                seats: "5",
                cost: 5000
            },
            loading: false
        })
    })

    it('it should handle GET_CURRENT_SUBSCRIPTION_SUCCESS', () => {
        expect(subscriptionReducer({}, {
            type: types.GET_CURRENT_SUBSCRIPTION_SUCCESS,
            payload: {
                plan: "best",
                name: "Best",
                seats: "5",
                cost: 5000
            }
        })).toEqual({
       
            currentSubscription: {
                plan: "best",
                name: "Best",
                seats: "5",
                cost: 5000
            },
            loading: false
        })
    })

})