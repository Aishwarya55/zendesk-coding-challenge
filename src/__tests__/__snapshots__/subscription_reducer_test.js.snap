// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`subscription reducer should return initial state 1`] = `
Object {
  "costPreview": Object {
    "cost": "",
    "name": "",
    "plan": "",
    "seats": 0,
  },
  "currentSubscription": Object {
    "cost": "",
    "name": "Basic",
    "plan": "basic",
    "seats": 0,
  },
  "loading": false,
  "previousSubscription": Object {
    "cost": "",
    "name": "Basic",
    "plan": "basic",
    "seats": 0,
  },
}
`;
